packages/core/src/tools/taskQueue.ts

```diff
@@ -0,0 +1,58 @@
+import { ONE_SECOND } from './utils/timeUtils'
+import { requestIdleCallback } from './requestIdleCallback'
+
+/**
+ * Maximum amount of time to delay the tasks. We don't want to wait too long before running tasks,
+ * as it might hurt reliability (ex: if the user navigates away, we might lose the opportunity to
+ * send some data). We also don't want to run tasks too often, as it might hurt performance.
+ */
+const IDLE_CALLBACK_TIMEOUT = ONE_SECOND
+
+/**
+ * Minimum amount of time to run the task when a timeout is reached. We should not run tasks for
+ * too long as it will hurt performance, but we should still run some tasks to avoid postponing them
+ * forever.
+ *
+ * Rational: Running tasks for 30ms every second (IDLE_CALLBACK_TIMEOUT) should be acceptable.
+ */
+export const TIMED_OUT_TIME_REMAINING = 30
+
+export interface TaskQueue {
+  push(task: Task): void
+}
+
+type Task = () => void
+
+export function createTaskQueue(): TaskQueue {
+  const pendingTasks: Task[] = []
+
+  function run(deadline: IdleDeadline) {
+    let timeRemaining: () => number
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏
```suggestion
    let executionTimeRemaining: () => number
```

Reply:
I followed the native naming here. But I can rename it, no problem.